File: app/src/main/java/com/icerrate/popularmovies/provider/cloud/RetrofitCallback.java
package com.icerrate.popularmovies.provider.cloud;

import android.util.Log;

import com.icerrate.popularmovies.view.common.BaseCallback;

import java.io.IOException;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * Created by Ivan Cerrate
 */
public class RetrofitCallback<T> implements Callback<T>  {

    private static final String TAG = RetrofitCallback.class.getSimpleName();

    private BaseCallback<T> callback;

    public RetrofitCallback(BaseCallback<T> callback) {
        this.callback = callback;
    }

    @Override
    public void onResponse(Call<T> call, Response<T> response) {
        if (response.isSuccessful()) {
            callback.onSuccess(response.body());
        } else {
            String errorMessage = "Unknown Error";
            if (response.errorBody() != null) {
                try {
                    errorMessage = response.errorBody().string();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            callback.onFailure(errorMessage);
        }
    }

    @Override
    public void onFailure(Call<T> call, Throwable t) {
        Log.d(TAG, "onFailure: " + t.getMessage());
        callback.onFailure(t.getMessage());
    }

}

File: app/src/main/java/com/icerrate/popularmovies/provider/cloud/RetrofitCallback.kt
package com.icerrate.popularmovies.provider.cloud

import android.util.Log
import com.icerrate.popularmovies.view.common.BaseCallback
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.IOException
import java.net.UnknownHostException

/**
 * @author Ivan Cerrate.
 */
class RetrofitCallback<T>(private val callback: BaseCallback<T>) : Callback<T> {

    companion object {
        private val TAG = RetrofitCallback::class.java.simpleName
    }

    override fun onResponse(call: Call<T>, response: Response<T>) {
        if (response.isSuccessful) {
            callback.onSuccess(response.body()!!)
        } else {
            var errorMessage = "Unknown Error"
            response.errorBody()?.let { errorBody ->
                try {
                    errorMessage = errorBody.string()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
            callback.onFailure(errorMessage)
        }
    }

    override fun onFailure(call: Call<T>, t: Throwable) {
        Log.d(TAG, "onFailure: ${t.message}")
        when (t) {
            is UnknownHostException -> callback.onFailure("No internet connection")
            else -> callback.onFailure(t.message ?: "Unknown error")
        }
    }
}